### ISSUE TITLE
Train effect doesn't perform well in my data set

### ISSUE THREAD

**Issue by ChangChunHe** [libAtoms/QUIP#157]
Well, I used `gap_fit` to train a data from VASP, 
```
gap_fit at_file=train.xyz e0=-0.34869725 \
gap={soap atom_sigma=0.1 l_max=12 n_max=12 cutoff=7.0 cutoff_transition_width=1.0 delta=0.01 add_species n_species=1 species_z={{ 5 }} n_sparse=2000 covariance_type=dot_product sparse_method=cur_points zeta=4.0} \
default_sigma={0.0001 0.002 1.0 1.0} sparse_jitter=1e-10 virial_parameter_name=none gp_file=train.xml
```
the  train.xyz  file is just like this:
```
19
config_type=crystal energy=-117.69129 Lattice="3.400000095600001 5.8889729108 0.0 -6.8000001907000005 2.9444864554 0.0 0.0 0.0 16.0" Properties=species:S:1:pos
:R:3:force:R:3
B  0.15 0.2 0.5 0.225327 -0.320459 0.0
B  0.3 0.4 0.5 -0.14864 -0.093425 0.0
B  0.4500000000000001 0.6000000000000001 0.5 -0.166485 0.252135 0.0
B  0.6000000000000001 0.8 0.5 0.116824 0.091986 0.0
B  0.75 0.0 0.5 0.220857 0.600541 -0.0
B  0.9000000000000001 0.2000000000000002 0.5 0.589107 -0.251245 0.0
B  0.05 0.4000000000000001 0.5 0.144987 -0.177133 -0.0
B  0.2000000000000002 0.6000000000000001 0.5 -0.007848 -0.392348 -0.0
B  0.3500000000000001 0.8 0.5 -0.121544 -0.096484 0.0
B  0.5 0.0 0.5 -0.0 0.0 0.0
B  0.6500000000000001 0.2000000000000002 0.5 0.121544 0.096484 0.0
B  0.8000000000000003 0.4000000000000004 0.5 0.007848 0.392348 0.0
B  0.9500000000000002 0.6000000000000001 0.5 -0.144987 0.177133 0.0
B  0.1000000000000001 0.8000000000000003 0.5 -0.589107 0.251245 -0.0
B  0.25 0.0 0.5 -0.220857 -0.600541 0.0
B  0.4000000000000004 0.2000000000000002 0.5 -0.116824 -0.091986 0.0
B  0.5500000000000003 0.4000000000000004 0.5 0.166485 -0.252135 0.0
B  0.7000000000000002 0.6000000000000001 0.5 0.14864 0.093425 0.0
B  0.8500000000000001 0.8000000000000003 0.5 -0.225327 0.320459 0.0
18
config_type=crystal energy=-113.76377 Lattice="3.400000095600001 5.8889729108 0.0 -6.8000001907000005 2.9444864554 0.0 0.0 0.0 16.0" Properties=species:S:1:pos
:R:3:force:R:3
B  0.15 0.2 0.5 0.3007 -0.111609 0.0
B  0.3 0.4 0.5 -0.146663 0.434729 0.0
B  0.4500000000000001 0.6000000000000001 0.5 0.715594 -0.157901 0.0
B  0.6000000000000001 0.8 0.5 0.609064 0.053636 0.0
B  0.75 0.0 0.5 0.190899 0.231007 0.0
B  0.9000000000000001 0.2000000000000002 0.5 0.200287 -0.185694 0.0
B  0.05 0.4000000000000001 0.5 0.276262 -0.241488 0.0
B  0.2000000000000002 0.6000000000000001 0.5 0.342396 0.095607 0.0
B  0.3500000000000001 0.8 0.5 -0.342396 -0.095607 0.0
B  0.5 0.0 0.5 -0.276262 0.241488 0.0
B  0.6500000000000001 0.2000000000000002 0.5 -0.200287 0.185694 0.0
B  0.8000000000000003 0.4000000000000004 0.5 -0.190899 -0.231007 0.0
B  0.9500000000000002 0.6000000000000001 0.5 -0.609064 -0.053636 0.0
B  0.1000000000000001 0.8000000000000003 0.5 -0.715594 0.157901 0.0
B  0.25 0.0 0.5 0.146663 -0.434729 0.0
B  0.4000000000000004 0.2000000000000002 0.5 -0.3007 0.111609 0.0
B  0.7000000000000002 0.6000000000000001 0.5 0.427889 -0.442801 0.0
B  0.8500000000000001 0.8000000000000003 0.5 -0.427889 0.442801 0.0

```
`train.xyz` contains 100 configurations, the above is just the first two configurations, and after training  I used `quip` to get the energy based on ``train.xml`` , 
```
quip atoms_filename=test.xyz param_filename=train.xml E > test_out

```
the file `test.xyz` is just the file ``train.xyz`` without `energy`, because I just want to know the fitting effect of `gap_fit` command, so below is the first two configurations of `test.xyz`,
```
19
config_type=crystal Lattice="3.400000095600001 5.8889729108 0.0 -6.8000001907000005 2.9444864554 0.0 0.0 0.0 16.0" Properties=species:S:1:pos:R:3
B 0.15 0.2 0.5
B 0.3 0.4 0.5
B 0.4500000000000001 0.6000000000000001 0.5
B 0.6000000000000001 0.8 0.5
B 0.75 0.0 0.5
B 0.9000000000000001 0.2000000000000002 0.5
B 0.05 0.4000000000000001 0.5
B 0.2000000000000002 0.6000000000000001 0.5
B 0.3500000000000001 0.8 0.5
B 0.5 0.0 0.5
B 0.6500000000000001 0.2000000000000002 0.5
B 0.8000000000000003 0.4000000000000004 0.5
B 0.9500000000000002 0.6000000000000001 0.5
B 0.1000000000000001 0.8000000000000003 0.5
B 0.25 0.0 0.5
B 0.4000000000000004 0.2000000000000002 0.5
B 0.5500000000000003 0.4000000000000004 0.5
B 0.7000000000000002 0.6000000000000001 0.5
B 0.8500000000000001 0.8000000000000003 0.5
18
config_type=crystal Lattice="3.400000095600001 5.8889729108 0.0 -6.8000001907000005 2.9444864554 0.0 0.0 0.0 16.0" Properties=species:S:1:pos:R:3
B 0.15 0.2 0.5
B 0.3 0.4 0.5
B 0.4500000000000001 0.6000000000000001 0.5
B 0.6000000000000001 0.8 0.5
B 0.75 0.0 0.5
B 0.9000000000000001 0.2000000000000002 0.5
B 0.05 0.4000000000000001 0.5
B 0.2000000000000002 0.6000000000000001 0.5
B 0.3500000000000001 0.8 0.5
B 0.5 0.0 0.5
B 0.6500000000000001 0.2000000000000002 0.5
B 0.8000000000000003 0.4000000000000004 0.5
B 0.9500000000000002 0.6000000000000001 0.5
B 0.1000000000000001 0.8000000000000003 0.5
B 0.25 0.0 0.5
B 0.4000000000000004 0.2000000000000002 0.5
B 0.7000000000000002 0.6000000000000001 0.5
B 0.8500000000000001 0.8000000000000003 0.5

```
but the predicted energy from `quip`  has rather error against the energy from `VASP`.

Here I plot a figure ,
![effect](https://user-images.githubusercontent.com/33366949/66483377-87e04180-ead7-11e9-8dd7-a3e3fb59dd25.png)

so I want to know whether there are some mistake in my parameters setting ?

**Comment by gabor1** [comment]
I have some comments on your parameters. I think your atom_sigma is too small. Potential energy surfaces are smooth, and I think an atom_sigma of 0.5 is more appropriate, maybe even 0.7. Second, 100 training configurations is very few. You are getting some good correlation in your plot, but you can't expect high accuracy from so little data - look at our papers, we train with a few thousand input configurations. Your default_sigma values are also very small - it's basically saying you expect to get a fitting error of 0.1 meV, which is too small - especially with 100 training configs, but even if you had 1000 I would say the best you can hope for is 0.001 eV. Similarly the force default_sigma is very small, I would try 0.2 ! Your delta is supposed to be the expected magnitiude of the atomic energy, so for fitting atomisation energies I would use about 2 eV.  (There is some relationship between delta and default_sigma, so the fact that all of them are too small is consistent, but remember these numbers cannot be just simply scaled because some of them have different units!!). You can make the fit faster by dropping the l_max quite a bit, it won't affect your accuracy very much. 

-- Gábor

**Comment by ChangChunHe** [comment]
OK, thanks for your reply, I used 100 trainning configuraions because I just wanted to try a trainning with too much time. I will try  another trainning with your advice.

**Comment by gabor1** [comment]
That’s of course a very good idea. Just don’t expect the results to be super accurate!

> On 10 Oct 2019, at 13:56, hecc <notifications@github.com> wrote:
> 
> OK, thanks for your reply, I used 100 trainning configuraions because I just wanted to try a trainning without too many time. I will try another trainning with your advice.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 


-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory
Pembroke College
University of Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by ChangChunHe** [comment]
But I used SOAP kernel with kernel ridge regression( poly kernel  will perform quite well), just based on several hundreds trainning configurations, the MAE of trainning  set will be quite small... If we want to know the details, I can supply the data and code.

**Comment by gabor1** [comment]
"But I used SOAP kernel with kernel ridge regress”

what are you talking about? your own independent KRR, or a gap_fit run ? if you simply take any kernel method, you can easily get arbitrarily small MAE on the training set by overfitting terribly. that doesn’t give you a usable potential...

> On 10 Oct 2019, at 14:06, hecc <notifications@github.com> wrote:
> 
> But I used SOAP kernel with kernel ridge regress, just based on several hundreds trainning configurations, the MAE of trainning set will be quite small...
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 


-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory
Pembroke College
University of Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by ChangChunHe** [comment]
Yes, as you say, I got a terribly overfitting result, so I have been thinking why I got the overfitting results, and your ``gap_fit`` can overcome overfitting problems? why? So can you tell me where should I read in your papers? Thanks a lot...

**Comment by gabor1** [comment]
> On 10 Oct 2019, at 14:18, hecc <notifications@github.com> wrote:
> 
> Yes, as you say, I got a terribly overfitting result, so I have been thinking why I got the overfitting results,

probably because you chose a small regularisation parameters to get a very good training error at the expense of a terrible test set error

> and your gap_fit can overcome overfitting problems? why?

it’s not gap_fit in general, but following my advice for how to select parameters (delta, sigma). also, including forces in the data leads to much less overfitting. 

> So can you tell me where should I read in your papers? Thanks a lot…

we provide parameters in all our papers, and the heuristics we use to get them - we have not written explicitly about the overfitting problem, because we don’t overfit! I am generally not a fan of optimising hyperparameters in some automated way, I think the risk of overfitting is too much. I’d much rather eliminate hyperparameters, or have settings that are natural and fixed. 


-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory
Pembroke College
University of Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by ChangChunHe** [comment]
well, thanks for your reply, of course, I also agree with your opinion that it is better if the number of  hyperparameters is less, too many parameters will cause overfitting preoblems. BTW,  I wonder whether  your code supplies the API of descriptors, that is, if I want to use myself-defined descriptors, how can I do?

**Comment by gabor1** [comment]
> On 10 Oct 2019, at 14:54, hecc <notifications@github.com> wrote:
> 
> well, thanks for your reply, of course, I also agree with your opinion that it is better if the number of hyperparameters is less, too many parameters will cause overfitting preoblems. BTW, I wander whether your code supplies the API of descriptors, that is, if I want to use myself-defined descriptors, how can I do?

quippy (the python interface) does provide the SOAP descriptors (and their derivatives!). 

gap_fit works with very many descriptors. we currently don’t have a documentation for this, but you can look at the list at the top of descriptors.f95 file. 

there is no API for gap_fit to work with externally calculated descriptors. 

> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 


-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory
Pembroke College
University of Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by ChangChunHe** [comment]
Oh, got it, thanks from my heart ==
